server/oauth_migration.go

```diff
@@ -0,0 +1,116 @@
+package main
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/mattermost/mattermost-plugin-api/cluster"
+	"github.com/pkg/errors"
+)
+
+const (
+	oauthMigrationStoreKey = "oauth_migration"
+	oauthMigrationMutexKey = "oauth_migration_mutex"
+
+	oauthMigrationStatusInProgress = "IN_PROGRESS"
+	oauthMigrationStatusComplete   = "COMPLETE"
+	oauthMigrationStatusError      = "ERROR"
+)
+
+func (p *Plugin) checkAndPerformOAuthTokenMigration() error {
+	mutexAPI := cluster.MutexPluginAPI(p.API)
+	mutex, err := cluster.NewMutex(mutexAPI, oauthMigrationMutexKey)
+	if err != nil {
+		return err
+	}
+
+	mutex.Lock()
+	defer mutex.Unlock()
+
+	var status string
+	err = p.client.KV.Get(oauthMigrationStoreKey, &status)
+	if err != nil {
+		return err
+	}
+
+	// Migration is in progress or already completed
+	if status != "" {
+		return nil
+	}
+
+	_, err = p.client.KV.Set(oauthMigrationStoreKey, oauthMigrationStatusInProgress)
+	if err != nil {
+		p.client.Log.Warn("error setting migration status to "+oauthMigrationStatusInProgress, "error", err.Error())
+	}
+
+	err = p.notifyAllConnectedUsersToReconnect()
+	if err != nil {
+		_, kvErr := p.client.KV.Set(oauthMigrationStoreKey, oauthMigrationStatusError)
+		if kvErr != nil {
+			p.client.Log.Warn("error setting migration status to "+oauthMigrationStatusError, "error", err.Error())
+		}
+
+		return err
+	}
+
+	_, err = p.client.KV.Set(oauthMigrationStoreKey, oauthMigrationStatusComplete)
+	if err != nil {
+		p.client.Log.Warn("error setting migration status to "+oauthMigrationStatusComplete, "error", err.Error())
+	}
+
+	return nil
+}
+
+func (p *Plugin) notifyAllConnectedUsersToReconnect() error {
+	allKeys := []string{}
+	page := 0
+	for {
+		keys, err := p.client.KV.ListKeys(page, 100)
+		if err != nil {
+			return errors.Wrap(err, "error listing keys for connected users")
+		}
+
+		if len(keys) == 0 {
+			break
+		}
+
+		keysToAdd := []string{}
+		for _, key := range keys {
+			if strings.HasSuffix(key, GitlabTokenKey) {
+				keysToAdd = append(keysToAdd, key)
+			}
+		}
+
+		allKeys = append(allKeys, keysToAdd...)
+		page++
```

Comment:
(suggestion)(non-blocking) We may save an extra request that's always done (for page 1) if we break if `len(keys)<100`